File: trees_and_graphs/count_good_nodes_in_binary_tree.cpp
#include "treenode.hpp"

#include <gtest/gtest.h>

#include <algorithm>
#include <limits>

//! @brief Recursive solution helper function returns answer to original problem
//! @param[in] node     Pointer to current TreeNode
//! @param[in] maxSoFar Largest value between root and current node
//! @return Number of nodes that are good
static int dfs(TreeNode* node, int maxSoFar)
{
    if (node == nullptr)
    {
        return 0;
    }

    const int left  = dfs(node->left, std::max(maxSoFar, node->val));
    const int right = dfs(node->right, std::max(maxSoFar, node->val));
    int ans         = left + right;
    if (node->val >= maxSoFar)
    {
        ++ans;
    }

    return ans;
}

//! @brief Find number of good nodes (path has no nodes with greater value)
//! @param[in] root Pointer to root of binary tree
//! @return Number of nodes that are good
static int goodNodesRecursive(TreeNode* root)
{
    return dfs(root, std::numeric_limits<int>::min());
}

TEST(GoodNodesTest, SampleTest)
{
    TreeNode four_lhs {4};
    TreeNode four_rhs {4};
    TreeNode five {5};
    TreeNode eight {8};
    TreeNode eleven {11};

    const TreeNode one {1};
    const TreeNode two {2};
    const TreeNode seven {7};
    const TreeNode thirteen {13};
    
    five.left  = &four_lhs;
    five.right = &eight;
    
    four_lhs.left = &eleven;
    
    eleven.left  = &seven;
    eleven.right = &two;

    eight.left  = &thirteen;
    eight.right = &four_rhs;

    four_rhs.right = &one;

    EXPECT_EQ(4, goodNodesRecursive(&five));
}

File: trees_and_graphs/count_good_nodes_in_binary_tree.java
public int dfs(TreeNode node, int maxSoFar)
{
    if (node == null)
    {
        return 0;
    }

    int left = dfs(node.left, Math.max(maxSoFar, node.val));
    int right = dfs(node.right, Math.max(maxSoFar, node.val));
    int ans = left + right;
    if (node.val >= maxSoFar)
    {
        ans += 1;
    }

    return ans;
}

public int goodNodesRecursive(TreeNode root)
{
    return dfs(root, Integer.MIN_VALUE);
}

class Pair
{
    TreeNode node;
    int maxSoFar;

    Pair(TreeNode node, int maxSoFar)
    {
        this.node = node;
        this.maxSoFar = maxSoFar;
    }
}

public int goodNodesIterative(TreeNode root)
{
    if (root == null)
    {
        return 0;
    }

    Stack<Pair> stack = new Stack<>();
    stack.push(new Pair(root, Integer.MIN_VALUE));
    int ans = 0;

    while (!stack.empty())
    {
        Pair pair = stack.pop();
        TreeNode node = pair.node;
        int maxSoFar = pair.maxSoFar;
        if (node.val >= maxSoFar)
        {
            ans++;
        }

        if (node.left != null)
        {
            stack.push(new Pair(node.left, Math.max(maxSoFar, node.val)));
        }

        if (node.right != null)
        {
            stack.push(new Pair(node.right, Math.max(maxSoFar, node.val)));
        }
    }

    return ans;
}